### PR TITLE
Set forwarded info as separate tag rather than replace peer

### PR DESCRIPTION
**Pull request by tylerbenson**
This avoids the complexity of managing multiple levels of forwarding presented as a comma separated list.

Also captures additional headers: `forwarded`, `x-forwarded-proto`, `x-forwarded-host`

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java [hunk 1]
```diff
@@ -51,11 +51,23 @@ public AgentSpan onRequest(
...
+      }
+      if (context.getForwardedHost() != null) {
+        span.setTag(Tags.HTTP_FORWARDED_HOST, context.getForwardedHost());
+      }
+      if (context.getForwardedIp() != null) {
+        span.setTag(Tags.HTTP_FORWARDED_IP, context.getForwardedIp());
+      }
+      if (context.getForwardedPort() != null) {
+        span.setTag(Tags.HTTP_FORWARDED_PORT, context.getForwardedPort());
+      }
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:70]: Please store these in local variables

File: dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecoratorTest.groovy [hunk 2]
```diff
@@ -99,14 +102,27 @@ class HttpServerDecoratorTest extends ServerDecoratorTest {
...
+      1 * span.setTag(Tags.PEER_HOST_IPV6, "3ffe:1900:4545:3:200:f8ff:fe21:67cf")
     } else {
-      1 * span.setTag(Tags.PEER_HOST_IPV6, "0::1")
+      1 * span.setTag(Tags.HTTP_FORWARDED_IP, "0::1")
+    }
+    1 * span.setTag(Tags.HTTP_FORWARDED_PORT, "123")
+    if (conn) {
+      1 * span.setTag(Tags.PEER_PORT, 555)
     }
-    1 * span.setTag(Tags.PEER_PORT, "123")
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecoratorTest.groovy:109]: It's not clear this tests anything except for the way you wrote the code. Please attempt to satisfy the request [here](https://github.com/DataDog/dd-trace-java/pull/2677/files#r624210427) without changing this test. If you find that you can't, your test isn't testing what the code does but how it does it, which is extremely hostile even to trivial refactoring.

Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecoratorTest.groovy]: I changed the mocks to not care about the number of invocations.  I think that should resolve your concern.

Comment (dougqh) [reply dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecoratorTest.groovy]: Mocks feel like overkill for this.  Although - admittedly, I'm not really a fan of mocking libraries period.

Why don't we just check the tags via a getTag mechanism?  Tracking calls to setTag just seems odd to me.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java [hunk 3]
```diff
@@ -78,18 +95,15 @@ public AgentSpan onRequest(
...
     }
 
-    final String ip = forwarded != null || connection == null ? forwarded : peerHostIP(connection);
-    if (ip != null) {
-      if (ip.indexOf(':') > 0) {
-        span.setTag(Tags.PEER_HOST_IPV6, ip);
-      } else {
-        span.setTag(Tags.PEER_HOST_IPV4, ip);
+    if (connection != null) {
+      final String ip = peerHostIP(connection);
```
Comment (bantonsson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:99]: This was previously set to the `x-forwarded-for` if present, and now it always picks up the `peerHostIP` from the connection. I just feel that I'm missing some explanation as to why this change in behavior is ok/desirable.

File: dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecoratorTest.groovy (hunk 2 above)
Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecoratorTest.groovy]: I just don't like tests like these. 

* They make it hard to refactor things because they end up specifying the implementation, normalising refactoring leading to destructive - as opposed to purely additive - changes to tests. 
* Sometimes they inadvertently verify interactions not required of the system under test in order to provoke behaviour, so a refactor can occasionally require a full run of the tests to detect all test assumptions violated (Groovy does **not** help here, where compiler errors would be a boon). 
* It's not always clear what is being tested and what is being provoked, and given that virtually any refactoring will lead to modification of the test, this risks refactoring leading to loss of specification. 
* Worst of all, I have found one case in our test suite where the mocks tested _impossible scenarios_ by ignoring the semantics of the system under test - in this parallel universe concocted by mocks, a class's superclass could have been an interface, so a refactor to take advantage of the fact that a class's superclass is not an interface failed.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpServerDecorator.java:99]: As hinted in the description, there is some subtle differences in how some frameworks handle comma separated lists of values in the header. For example, `RemoteIpValve` is configurable and often filters out local (eg `10.x`, `192.168.x`) ip addresses. Rather than match that flexibility, I think it's better to just present all the information.

**Review by richardstartin — APPROVED** [review]

**Review by bantonsson — APPROVED** [review]
